Include/PGE/String/String.h

```diff
@@ -287,47 +290,68 @@ class String {
         bool isEmpty() const;
 
     private:
-        String(int size);
+        String(nullptr_t) { } // Dummy constructor
         String(const char* cstr, size_t size);
         String(const String& other, int from, int cnt);
 
-        static void copy(String& dst, const String& src);
-
-        static constexpr int SHORT_STR_CAPACITY = 16;
+        static constexpr int SHORT_STR_CAPACITY = 40;
 
         struct Data {
-            int strByteLength = -1;
-
-            int cCapacity = SHORT_STR_CAPACITY;
-
             // Lazily evaluated.
-            mutable u64 _hashCode;
+            mutable u64 _hashCode = 0;
             mutable int _strLength = -1;
-            mutable bool _hashCodeEvaluted = false;
+
+            int strByteLength = -1;
+        };
+
+        struct CoreInfo {
+            char* chs;
+            Data* data;
         };
 
         struct Shared {
             Data data;
+            int cCapacity;
             std::unique_ptr<char[]> chs;
+            const CoreInfo get() {
+                return { chs.get(), &data };
+            }
         };
 
         struct Unique {
             Data data;
             char chs[SHORT_STR_CAPACITY];
+            const CoreInfo get() {
+                return { chs, &data };
+            }
+        };
+
+        struct Literal {
+            std::variant<Data, Data*> data;
```

Comment:
Why can this hold pointers to Data? It seems like a small enough struct to consider holding as a value exclusively

Reply:
Literals share length and hash code globally, seems like the simplest way to accomplish that

Reply:
The data for literals is immutable, isn't it? Copying is probably safer

Reply:
Moreover this doesn't appear to ever get assigned anything that isn't a pointer

Reply:
By using a pointer if two String objects have the same literal and one of them gets updated the other one will as well.

Reply:
I'm not sure I understand. Are literals _not_ immutable? Why would we want these two strings to update together?

Reply:
They are, but their hashcode and length are still lazily evaluated.

Reply:
I see. In that case why would this need to be able to hold a value rather than a pointer?

Reply:
For the strByteLength which is not lazily evaluated

Reply:
I see no circumstance in which this is used this way though. `String::LiteralData::shareData()` always emplaces a pointer.

Reply:
https://github.com/juanjp600/pge/blob/a52f9013f108e91da7a76ec811c9c8f68f8afea6/Src/String/String.cpp#L255

Reply:
(And String::byteLength() for where it's used)

Reply:
The idea is to not have a map lookup in the literal constructor